### ISSUE TITLE
Support for ST_AsMVT and ST_AsMVTGeom?

### ISSUE THREAD

**Issue by mhkeller** [duckdb/duckdb_spatial#241]
In PostGIS, you can serve vector tiles using a combination of [`ST_AsMVT`](https://postgis.net/docs/ST_AsMVT.html) and [`ST_AsMVTGeom`](https://postgis.net/docs/ST_AsMVTGeom.html). It would be really neat if you could do something similar with duckdb. Is there any plan to put these on the roadmap for duckdb_spatial?

Here are a few examples of how those functions can be strung together to create a vector-tile server:

1. [minimal-mvt](https://github.com/pramsey/minimal-mvt/blob/58f3e695a305f42024dcf0ba395590bf39b0b573/minimal-mvt.py#L109)
2. [mvt-server](https://github.com/mhkeller/postgis-line-repro/blob/b49a91079c00592c66f934be932dff19c28179e4/routes/generate-vector-tile.sql#L5)

**Comment by Maxxen** [comment]
Yes! I have a sketch for these functions in an old PR so will definitely be added in the future!

**Comment by mhkeller** [comment]
That's great to hear. I'm looking forward to seeing it. Thanks for your work!

**Comment by talllevy** [comment]
that is awesome @Maxxen. mind sharing the PR or branch? I was just looking for this type of functionality and would love to try to help make it happen

**Comment by Maxxen** [comment]
@talllevy Hmm, looks like I didnt push it after all and I can't find it in my git stash... 

Anyway, the difficulty lies in writing the protobuf out incrementally in the aggregate function. An ideal implementation would just write out the bytes into a growable buffer similar to how e.g. most string aggregate functions already work, which would allow DuckDB to offload the aggregate states to disk. But a first step could also just be to keep a pointer in the aggregate state to a std::vector where we collect all the geometries and don't do the actual protobuf mvt encoding until the finalization step. I would use [vtzero](https://github.com/mapbox/vtzero) as we already use and bundle protozero, and duckdb core also provides a mini zip wrapper we can use to compress the final tile.

I should warn you though that this will probably only work well for individual/small collection of tiles and not for generating entire mbtile or pmtile archives as duckdbs current arbitrary grouped aggregate require every input to pass through every aggregate group, meaning you'll have to check intersection between every input geometry and every tile. There are smarter algorithms to do this by e.g. building a quad tree and/or recursively subdividing the tiles but that would probably require implementing a sorted/spatially aware aggregate operator a custom operator entirely (or multiple chained together), which I would like to look into at some point.

**Comment by talllevy** [comment]
thanks! I'll take a look and see if I can tackle this

**Comment by swdmike** [comment]
any update of ST_AsMVT, we really need this

**Comment by talllevy** [comment]
hey all. I've still been working on this, albeit very very slowly due to other obligations. I have been testing my implementation of ST_TileEnvelope and ST_MakeEnvelope first as a precursor to implementing the AsMVT functions. I will take a look at ST_Segmentize too. I'll dedicate more time to this over the next two weeks. Will give an update again then!

**Comment by Maxxen** [comment]
I've revisited this again the last couple of days, now that we have fast spatial filtering through the R-Tree index I think using duckdb as a tile server backend could become somewhat feasible. In progress branch is [here](https://github.com/Maxxen/duckdb_spatial/tree/stasmvt). Not the prettiest code I've ever written but the biggest issue so far seems to be getting `ST_AsMVTGeom` to work when GEOS rejects clipped geometries as invalid (which it created itself). Still also need to make `ST_AsMVT` handle non-geometry property columns and use the DuckDB memory allocator, but that shouldn't be that hard. I'll link to this issue once I get a PR up.

**Comment by majid-saeed** [comment]
@Maxxen any update here, anybody active on this?

**Comment by ilyabo** [comment]
> As a side note, does anybody have any tips on how to test/verify/visualize MVTs locally? QGIS won't accept them, so I'm just checking that ogrinfo can parse.

@Maxxen I've built this PoC with a map visualization, should be easy to adapt to `ST_AsMVT`:  https://github.com/ilyabo/duck-tiles/

**Comment by Maxxen** [comment]
Aiming to get this in for DuckDB v1.4!

**Comment by bartzwemmer** [comment]
> [@Maxxen](https://github.com/Maxxen) In which version, these functions be released?

In version 1.4.0 of DuckDB indeed, which is already available. Based on @Maxxen example, I made[ a blog post](https://onegiantleap.dev/duckdb-vector-tiles-demo/) and demo with the released version.
